scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -0,0 +1,133 @@
+
+(def TimePeriodMs NonNegInt)
+
+;; Elements missing:
+;;   1. Attributes (chip set, az, ..) -- have a combinitorial problem here
+(def HostInfo
+  {:count NonNegInt
+   :instance-type s/Str
+   :cpus PosNum
+   :mem PosNum
+   (s/optional-key :gpus) PosNum})
+
+(defprotocol HostFeed
+  "Protocol defining a service to get information on hosts that can be purchased."
+  (get-available-host-info [this] "Returns a list of host info maps conforming
+                                   to HostInfo schema above"))
+
+(def Schedule
+  {TimePeriodMs {:suggested-matches {HostInfo [s/Uuid]}}})
+
+(defprotocol Optimizer
+  "Protocol defining a tool to produce a schedule to execute"
+  (produce-schedule [this queue running available host-infos]
+                    "Returns a schedule of what jobs to place on what machines
+                     and what hosts to purchase at different time steps.
+                     Conforms to the Schedule schema above
+
+                     Parameters:
+                     queue -- Ordered list of jobs to run (see cook.mesos.schema job)
+                     running -- Set of tasks running (see cook.mesos.schema instance)
+                     available -- Set of offers outstanding
+                     host-infos -- Host infos from HostFeed
+                     
+                     Returns:
+                     A schedule: a map where the keys are milliseconds in the future 
+                     and  the values are recommendations to take at that point in the future"))
+
+(defn create-dummy-host-feed
+  "Returns an instance of HostFeed which returns an empty list of hosts"
+  [_]
+  (log/info "Creating dummy host feed")
+  (reify HostFeed
+    (get-available-host-info [this]
+      [])))
+
+(defn create-dummy-optimizer
+  "Returns an instance of Optimizer which returns an empty schedule"
+  [_]
+  (log/info "Creating dummy optimizer")
+  (reify Optimizer
+    (produce-schedule [this queue running available host-infos]
+      {0 {:suggested-matches {}}})))
+
+(defn optimizer-cycle!
+  "Starts a cycle that:
+   1. Gets queue, running, offer and purchasable host info
+   2. Calls the `optimizer` to get a schedule
+
+   Parameters:
+   get-queue -- fn, no args fn that returns ordered list of jobs to run
+   get-running -- fn, no args fn that returns a set of tasks running
+   get-offers -- fn, no args fn that returns a set of offers
+   host-feed -- instance of HostFeed
+   optimizer -- instance of Optimizer
+   interval -- joda-time period
+
+   Raises an exception if there was a problem in the execution"
+  [get-queue get-running get-offers host-feed optimizer]
+  (let [queue (future (get-queue))
+        running (future (get-running))
+        offers (future (get-offers))
+        host-infos (get-available-host-info host-feed)
+        _ (s/validate [HostInfo] host-infos)
+        schedule (produce-schedule optimizer @queue @running @offers host-infos)]
+    (s/validate Schedule schedule)))
```

Comment:
Should we publish the `schedule` somewhere for a potential consumer to be able to consume it? Or we can return the value so that the caller can decide how to use it.

Reply:
Yes, we should publish it, but I think it should be in a separate PR so this one doesn't get too big.